usr/share/rear/conf/Ubuntu.conf

```diff
@@ -1,2 +1,18 @@
 PROGS=( "${PROGS[@]}" initctl MAKEDEV )
 
+#####
+# TCG Opal 2 PBA system only: Include Plymouth graphical boot animation
+
+OPAL_PBA_PROGS+=( plymouth plymouthd )
+
+OPAL_PBA_COPY_AS_IS+=( /etc/alternatives/*plymouth* /usr/lib/x86_64-linux-gnu/plymouth /usr/share/plymouth /usr/share/fonts/truetype/ubuntu )
+
+# NOTE: 'build/GNU/Linux/100_copy_as_is.sh' auto-detects shared libraries referenced by executables from COPY_AS_IS
+# but fails to consider shared libraries (*.so) from COPY_AS_IS. Adding those to LIBS gets them covered.
```

Comment:
As its name tells COPY_AS_IS is rather meant to copy as is.

To copy shared libraries use the LIBS array.
To copy "nice to have" executables use the PROGS array.
To copy required executables use the REQUIRED_PROGS array.

For elements in the LIBS, PROGS, and REQUIRED_PROGS arrays
RequiredSharedObjects is called to copy also dependant shared objects.

For elements in the COPY_AS_IS array there is an exception to the
"only copy as is" rule that for executrables in the COPY_AS_IS array
RequiredSharedObjects is called to copy also dependant shared objects
so that for shared libraries in the COPY_AS_IS array that have the executable
permission bit set RequiredSharedObjects is called but not for libraries
in the COPY_AS_IS array that do not have the executable permission bit set.
The reason for that exception is that programs in the COPY_AS_IS array
are handled gracefully but the COPY_AS_IS array is not meant for libraries.

I think this all looks as if it has grown over the past into something that
is overcomplicated now and should perhaps be replaced by a single plain
RECOVERY_SYSTEM_INCLUDE array where for each element its dependant
things are automatically determined and included in the recovery system
(e.g. also needed interpreters for scripts and more advanced things like that)
but that is for some dreamlike future, cf. https://github.com/rear/rear/issues/1390

Reply:
In many cases, using `LIBS` would be straight forward. In this case, there was a directory tree containing a mix of shared libraries and other stuff, making it a good candidate for `COPY_AS_IS`. At some point, one gets used to all that auto-stuff luxury in ReaR (such as integrating dependencies) and takes it for granted. Then it becomes a usability problem where it doesn't work consistently across the bench as in this case. Normally, it might not be such a big problem and is easily detected afterwards. Here the error message was deeply buried in a chatty Plymoth debug log which was only available on a limited text screen without ssh access or easily accessible export capabilities.

Reply:
@OliverO2 
I am afraid I think the plain truth is that
there is not so much of consistency in ReaR and
not much admin usability and no end-user usability.

This is one of the reasons behind why I wrote the
"Inappropriate expectations" section in
https://en.opensuse.org/SDB:Disaster_Recovery
and therein also things like
```
... intended for experienced users and system admins.
There is no easy user-frontend ...
...
Experienced users and system admins
...
After an experienced admin had set it up
...
To get sufficient understanding how ReaR works
you need to use it yourself,
play around with it yourself,
do some trial an error experiments with it yourself
```
and
https://github.com/rear/rear/blob/master/usr/share/rear/conf/default.conf#L36

Even more I appreciate your endurance
to investigate such problems
until you got the root cause.

Reply:
Now I am wondering if there could be ever a use case
to have a shared object (library) in the ReaR recovery system
but not also all of its dependant required other shared objects?

I think a shared object (library) can not at all work
without its dependant required other shared objects
so I think
* either have a shared object with all its required other shared objects
* or do not have the shared object at all because it is useless

Or could a shared object be useful in any way
without all its required other shared objects?

Reply:
> Or could a shared object be useful in any way
> without all its required other shared objects?

No. It just won't load. So it is a non-functional trap waiting for someone to step into. ;-)

And I'd like to object, as least a bit: While ReaR is not that high on (even developer-oriented) usability, it is not that bad: ReaR has so many little functions included that try to auto-detect and/or auto-correct stuff which would be hard to get right the first time. So it just works in many cases, which makes it get usability points. ReaR also includes considerable effort to spot problems and provide useful information in logs, which makes it get additional points.

So maybe improving `COPY_AS_IS` to inspect shared libraries would be a good idea. Currently, it lets `ldd` look at every shell/python/whatever executable script, so why not make it inspect shared libraries? The only problem would be the naming. It's not just `*.so` or `*.so*` (as this would include `syslog.socket`) or `lib*.so*`. You could have:
* `libSegFault.so`
* `libacl.so.1.1.0`
* `pam_gdm.so`

Maybe something like `*.so` plus `*.so.[0-9]*` would do...

Reply:
My reason behinf my question in
https://github.com/rear/rear/pull/2278#discussion_r347445136
is exactly what you propose in your reply
https://github.com/rear/rear/pull/2278#discussion_r347458656
```
improving COPY_AS_IS to inspect shared libraries would be a good idea
```
but I would try to go even further and try to
improve COPY_AS_IS to inspect all regular files
because as far as I know `ldd` won't run amok
for files that it cannot work with
e.g. on my openSUSE Leap 15.0 system
```
# ldd /etc/fstab 
ldd: warning: you do not have execution permission for `/etc/fstab'
        not a dynamic executable

# ldd /boot/vmlinuz-4.12.14-lp150.12.82-default
ldd: warning: you do not have execution permission for `/boot/vmlinuz-4.12.14-lp150.12.82-default'
        not a dynamic executable

# ldd /dev/sda
ldd: /dev/sda: not regular file
```
which may result in the end to inspect all regular files
in the recovery system, i.e. to enhance first and foremost
usr/share/rear/build/default/990_verify_rootfs.sh
to check all files if they have all required libraries
actually accessible inside the recovery system.

Reply:
All files would probaby be a bit too much. I've tested a bit and found that this one would get me all shared libraries:
```bash
find / -mount \( -name \*.so -o -name \*.so.[0-9]\* \) -print
```

But some caution from http://man7.org/linux/man-pages/man1/ldd.1.html seems advisable:

> Security
>
> Be aware that in some circumstances (e.g., where the program speci‐
> fies an ELF interpreter other than ld-linux.so), some versions of ldd
> may attempt to obtain the dependency information by attempting to
> directly execute the program, which may lead to the execution of
> whatever code is defined in the program's ELF interpreter, and per‐
> haps to execution of the program itself.  (In glibc versions before
> 2.27, the upstream ldd implementation did this for example, although
> most distributions provided a modified version that did not.)
>
> Thus, you should never employ ldd on an untrusted executable, since
> this may result in the execution of arbitrary code.  A safer alterna‐
> tive when dealing with untrusted executables is:
>
>    $ objdump -p /path/to/program | grep NEEDED
>
> Note, however, that this alternative shows only the direct dependen‐
> cies of the executable, while ldd shows the entire dependency tree of
> the executable.

So if someone configured ReaR to include user-writable directories in `COPY_AS_IS`, executables lying around there might be executed with root privileges. This could already happen currently, so it would probably be a good idea to limit executable examination to safe paths only.

It is not clear from the above statement whether a shared library would be considered an executable, too (in most respects, it is). So if `COPY_AS_IS` were to be extended, it would probably be safer to limit automatic examination of shared libraries to safe paths only (such as `/lib` and `/usr/lib`) as well.

I still think that extending `COPY_AS_IS` would be a good idea and might even save you some support headaches later on.

Reply:
I was thinking about if it could be sufficient
to run `ldd` only for the regular files in some
"well known" directories in the recovery system.

I did a quick test to see in which directories in the recovery system
`ldd` found regular files that are a dynamic executable
i.e. in which directories in the recovery system there are
files where running `ldd` makes sense:
```
# ROOTFS_DIR=/tmp/rear.c3mdOvGKSRIkH8L/rootfs/

# for d in $( for f in $( find $ROOTFS_DIR -type f -printf '/%P\n' ) ; \
 do ldd $f 2>/dev/null | grep -v 'not a dynamic executable' \
 && echo $f is a dynamic executable ; done \
 | grep 'is a dynamic executable' | cut -d ' ' -f1 ) ; \
 do dirname $d ; done | sort -u

/bin
/lib64
/usr/lib/dracut
/usr/lib/dracut/modules.d/99kdump
/usr/lib/perl5/5.26.1/x86_64-linux-thread-multi/CORE
/usr/lib/systemd
/usr/lib/systemd/system-generators
/usr/lib/udev
/usr/lib64
/usr/lib64/rsyslog
```

Accordingly files where running `ldd` makes sense
can appear in arbitrary directories in the recovery system
so it is not sufficient to run `ldd` only for the regular files
in some "well known" directories in the recovery system.

Reply:
@jsmeix GitHub refused to accept log attachments here (probably wisely so), so I'm continuing this in a separate discussion issue #2279.